File: core/src/main/java/dev/vibris/core/ShaderLogSink.java
package dev.vibris.core;

import dev.vibris.api.ReloadResult;
import dev.vibris.protocol.v1.ArtifactMetadata;

import java.io.IOException;
import java.util.List;

interface ShaderLogSink {
    ArtifactMetadata writeShaderLog(
        String workspaceId,
        String requestId,
        List<ReloadResult.Diagnostic> diagnostics
    ) throws IOException;

    static ShaderLogSink none() {
        return (workspaceId, requestId, diagnostics) -> {
            throw new IOException("Shader log storage is unavailable.");
        };
    }
}

File: core/src/main/kotlin/dev/vibris/core/ShaderLogSink.kt
package dev.vibris.core

import dev.vibris.api.ReloadResult
import dev.vibris.protocol.v2.ArtifactMetadata
import java.io.IOException

internal fun interface ShaderLogSink {
    @Throws(IOException::class)
    fun writeShaderLog(
        workspaceId: String,
        requestId: String,
        diagnostics: List<ReloadResult.Diagnostic>,
    ): ArtifactMetadata

    companion object {
        @JvmStatic
        fun none(): ShaderLogSink =
            ShaderLogSink { _, _, _ ->
                throw IOException("Shader log storage is unavailable.")
            }
    }
}